### DISCUSSION TITLE
Play application behind load balancer

### DISCUSSION THREAD

**Discussion by szymond** [playframework/playframework#11562] [Q&A]
Hey,
I am considering running Play application behind load balancer. 

1. Do I understand correctly that in this case, each deployment should have same application key? 

2. I am running application in "public" cloud. I assume traffic can be sniffed there. So I wish to encrypt traffic between load balancer and applications. I think for such case, selfsigned certificate should suffice. But when I let Play generate by itself a self-signed certificate, application logs warnings that such certificate should not be used on production. Is there reason for that or is my use case is not considered valid production case?

**Comment by Squeng** [comment]
1. Yes, each deployment should have the same application key such that when the session cookie and CSRF token are signed by one deployment, the signature can be verified by another deployment.
2. As far as the user/browser is concerned, a self-signed certificate cannot be differentiated from a certificate by an attacker unless you disseminate your certificate via an alternative, trusted (and hopefully trustworthy) channel. Anyway, these days, public clouds (such as my provider Clever Cloud) provide free certificates that are properly rooted in a user's browser thanks to Let's Encrypt, which is even easier to use than using a (self-signed) certificate within the Play app: https://www.clever-cloud.com/doc/administrate/ssl/

**Comment by szymond** [comment]
Thanks. Regarding certificate question, you might have not understood it (or just used opportunity to promote your provider). The described scenario does not involve user/browser. It's about communication between LB and application. So no browser is involved, users' secure connection is terminated on LB (and there I can have Let's Encrypt certificate).

**Reply by Squeng** [reply]
(I'm just a Clever Cloud customer, not affiliated with them in any way.)

It depends on both the cloud provider (i.e., how well it isolates various apps/customers) and the requirements (e.g., "all communication has to be encrypted"). For example, in case of a banking application running in a private cloud, we had the requirement to also use SSL/TLS "behind" the load balancer (or gateway in our case). There we generated the necessary certificates and installed them both in the gateway as well as the Play app and configured the Play app according to https://www.playframework.com/documentation/2.8.x/ConfiguringHttps.
